src/client/game.cpp

```diff
@@ -3730,15 +3731,56 @@ bool Game::nodePlacement(const ItemDefinition &selected_def,
 		LocalPlayer *player = client->getEnv().getLocalPlayer();
 
 		// Don't place node when player would be inside new node
-		// NOTE: This is to be eventually implemented by a mod as client-side Lua
-		if (!predicted_f.walkable ||
+		const bool skip_check =  !predicted_f.walkable ||
 				g_settings->getBool("enable_build_where_you_stand") ||
-				(client->checkPrivilege("noclip") && g_settings->getBool("noclip")) ||
-				(predicted_f.walkable &&
-					neighborpos != player->getStandingNodePos() + v3s16(0, 1, 0) &&
-					neighborpos != player->getStandingNodePos() + v3s16(0, 2, 0))) {
+				(client->checkPrivilege("noclip") && g_settings->getBool("noclip"));
+
+		// We are simply colliding the player's current collisionbox in the world to a
+		// "normal node" collision box. This can be improved to take nodeboxes.
+		bool narrow_phase_collision = false;
+		if (!skip_check) {
+			// Collect all node collision boxes.
+			auto cboxes = new std::vector<aabb3f>();
+			predicted_node.getCollisionBoxes(nodedef_manager, cboxes, 0);
+
+			// Convert the node placement prediction position into BS space.
+			const v3f BS_space_p = v3f(
+				(f32)p.X * BS,
+				(f32)p.Y * BS,
+				(f32)p.Z * BS
+			);
+
+			// Create the player's collision box in BS real world space.
+			aabb3f cbox = player->getCollisionbox();
+			v3f pos = player->getPosition();
+			aabb3f player_cbox(cbox.MinEdge + pos,cbox.MaxEdge + pos);
+			for (uint i = 0; i < cboxes->size(); i++) {
+				// Now convert the collision box into BS real world space. 
+				const aabb3f rawCbox = cboxes->at(i);
+				const aabb3f normal_node_cbox(
+					rawCbox.MinEdge + BS_space_p,
+					rawCbox.MaxEdge + BS_space_p
+				);
+
+				// Now collide it. Narrow phase is done.
+				narrow_phase_collision = player_cbox.intersectsWithBox(normal_node_cbox);
+				if (narrow_phase_collision)
+					break;
+			}
+			// Release the memory.
+			free(cboxes);
```

Comment:
Ack--BAD!  Don't use `free()` on something allocated with `new`!  That's major undefined behavior.

Better yet, do `std::vector<aabb3f> cboxes;` above if possible; avoid heap allocation whenever possible, especially raw allocation with `new`.

Reply:
You should never* use `new`, `delete`, `malloc`, or `free` in modern C++ - use smart pointers (std::unique_ptr) or stack allocation instead

(* - well, only if you're writing a low level data structure class that encapsulates these things)

Reply:
Fixed in: 30e50ea2c4759dbdc59a93f98eb21239daf7e56c